oil_lang/expr_eval.py

```diff
@@ -176,10 +183,156 @@ def _ValueToPyObj(val):
       val = cast(value.Dict, UP_val)
       return val.d
 
+    elif case(value_e.Eggex):
+      val = cast(value.Eggex, UP_val)
+      return objects.Regex(val.expr)
+
     else:
       raise NotImplementedError()
+
+
+def _ValuesEqual(left, right):
+  # type: (value_t, value_t) -> bool
+  if left.tag() != right.tag():
+    return False
+
+  UP_left = left
+  UP_right = right
+  with tagswitch(left) as case:
+    if case(value_e.Undef):
+      return True
+
+    elif case(value_e.Bool):
+      left = cast(value.Bool, UP_left)
+      right = cast(value.Bool, UP_right)
+      return left.b == right.b
+
+    elif case(value_e.Int):
+      left = cast(value.Int, UP_left)
+      right = cast(value.Int, UP_right)
+      return left.i == right.i
+
+    elif case(value_e.Float):
+      left = cast(value.Float, UP_left)
+      right = cast(value.Float, UP_right)
+      return left.f == right.f
+
+    elif case(value_e.Str):
+      left = cast(value.Str, UP_left)
+      right = cast(value.Str, UP_right)
+      return left.s == right.s
+
+    elif case(value_e.MaybeStrArray):
+      left = cast(value.MaybeStrArray, UP_left)
+      right = cast(value.MaybeStrArray, UP_right)
+      if len(left.strs) != len(right.strs):
+        return False
+
+      for i in xrange(0, len(left.strs)):
+        if left.strs[i] != right.strs[i]:
+          return False
+
+      return True
+
+    elif case(value_e.List):
+      left = cast(value.List, UP_left)
+      right = cast(value.List, UP_right)
+      if len(left.items) != len(right.items):
+        return False
+
+      for i in xrange(0, len(left.items)):
+        if not _ValuesEqual(left.items[i], right.items[i]):
+          return False
+
+      return True
+
+    elif case(value_e.AssocArray):
+      left = cast(value.Dict, UP_left)
+      right = cast(value.Dict, UP_right)
+      if len(left.d) != len(right.d):
+        return False
+
+      for k in left.d.keys():
+        if not k in right.d or right.d[k] != left.d[k]:
```

Comment:
style nitpick: in Python you can use `k not in right.d`

Reply:
ah great. wasn't sure if mycpp could deal with `not in`